### PR TITLE
FY sampling procedure

### PR DESCRIPTION
**Pull request by EnricaBelfiore**
Add correlations to independent fission yield covariance matrix:
---

This notebook shows the perturbation process to obtain an Endf-6 file with the perturbed IFY and a visual comparison between the correlation matrices obtained from the prior covariance matrix and the updated one is shown.

### CODE REVIEW

**Comment by luca-fiorito-11** [comment]
Does this work without merging PR #126 ?

**Comment by EnricaBelfiore** [comment]
The aim of the last commit is to correct the errors already present in the `develop` branch. The Notebook inserted will work with these corrections. The methods in the PR #126 will be inserted if needed when they will be properly tested in other Notebooks (ongoing)

**Comment by AitorBengoechea** [comment]
For version control, this should go on the pull request https://github.com/AitorBengoechea/sandy/tree/fy

**Comment by AitorBengoechea** [comment]
For version control, this should go on the pull request https://github.com/AitorBengoechea/sandy/tree/fy

**Comment by EnricaBelfiore** [comment]
> For version control, this should go on the pull request https://github.com/AitorBengoechea/sandy/tree/fy

Not really, all the methods used in the notebook are reported in this PR. There is still an error in the implementation of the `sandy.gls.constrained_gls_update()` method which leads to wrong results in some tests used in FY. It will be fixed in another PR.

File: sandy/fy.py [hunk 1]
```diff
@@ -328,11 +390,11 @@ def get_mass_yield_sensitivity(self):
             aggfunc="sum"
             ).fillna(0)
         return mass_yield_sensitivity
-
-    def custom_perturbation(self, zam, mt, e, zap, pert):
+    
+    def custom_perturbation(self, pert, **kwargs):
```
Comment (luca-fiorito-11) [inline sandy/fy.py:394]: this has to be rediscussed

File: sandy/fy.py [hunk 2]
```diff
@@ -405,22 +503,22 @@ def apply_bmatrix(self, zam, e, decay_data, keep_fy_index=False):
         >>> npfy = Fy(minimal_fytest_2)
         >>> npfy_pert = npfy.apply_bmatrix(942390, 5.00000e+05, decay_fytest)
         >>> npfy_pert.data[npfy_pert.data.MT == 454]
-        	 MAT	 MT	   ZAM	   ZAP	          E	         FY	        DFY
-        3	9437	454	942390	591480	5.00000e+05	8.00000e-01	4.00000e-02
-        4	9437	454	942390	591481	5.00000e+05	1.00000e+00	5.00000e-02
-        5	9437	454	942390	601480	5.00000e+05	-1.60000e+00	1.00000e-01
-        6	9437	454	942390	621480	5.00000e+05	-2.00000e-01	1.00000e-02
+            MAT   MT     ZAM     ZAP           E           FY         DFY
```
Comment (luca-fiorito-11) [inline sandy/fy.py:506]: We will need a notebook to check this...and also maybe to clean up the coding

File: sandy/fy.py [hunk 3]
```diff
@@ -497,22 +595,22 @@ def apply_qmatrix(self, zam, energy, decay_data, keep_fy_index=False):
         >>> npfy = Fy(minimal_fytest_2)
         >>> npfy_pert = npfy.apply_qmatrix(942390, 5.00000e+05, decay_fytest)
         >>> npfy_pert.data[npfy_pert.data.MT == 459]
-             MAT	 MT	   ZAM	   ZAP	          E	         FY 	    DFY
-        3	9437	459	942390	591480	5.00000e+05	1.00000e-01	4.00000e-02
-        4	9437	459	942390	591481	5.00000e+05	2.00000e-01	5.00000e-02
-        5	9437	459	942390	601480	5.00000e+05	6.00000e-01	1.00000e-01
-        6	9437	459	942390	621480	5.00000e+05	6.00000e-01	1.00000e-01
+             MAT   MT     ZAM     ZAP           E          FY         DFY
```
Comment (luca-fiorito-11) [inline sandy/fy.py:598]: Same...notebook for this, but in another PR

File: sandy/fy.py [hunk 4]
```diff
@@ -593,84 +685,92 @@ def gls_cov_update(self, zam, e, Vy_extra=None,
...
-        601480	-4.00000e-03	-5.00000e-03	 9.00000e-03
+        591480   9.90476e-04 -9.52381e-04 -3.80952e-05
+        591481  -9.52381e-04  1.01190e-03 -5.95238e-05
+        601480  -3.80952e-05 -5.95238e-05  9.76190e-05
         """
         # Divide the data type:
         conditions = {'ZAM': zam, "E": e}
         fy_data = self._filters(conditions).data\
                       .set_index('ZAP')[['MT', 'DFY']]
         Vx_prior = fy_data.query('MT==454').DFY
```
Comment (luca-fiorito-11) [inline sandy/fy.py:750]: this is not Vx_prior...coding to be adjusted later

File: sandy/fy.py [hunk 5]
```diff
@@ -544,14 +642,14 @@ def apply_qmatrix(self, zam, energy, decay_data, keep_fy_index=False):
         data = pd.concat([data, calc_values], ignore_index=True)
         return self.__class__(data)
 
-    def gls_cov_update(self, zam, e, Vy_extra=None,
-                       kind='mass yield', decay_data=None, threshold=None):
+    def gls_cov_update(self, zam, e, S, Vy_extra=None,
+                      rows=None, threshold=None):
```
Comment (luca-fiorito-11) [inline sandy/fy.py:646]: Use of `threshold` to be discussed

File: sandy/fy.py [hunk 6]
```diff
@@ -775,9 +888,7 @@ def gls_update(self, zam, e, y_extra=None, Vy_extra=None,
         data = pd.concat([data, calc_values], ignore_index=True)
         return self.__class__(data)
 
-    def ishikawa_factor(self, zam, e, Vy_extra,
-                        kind='mass yield', decay_data=None,
-                        threshold=None):
+    def ishikawa_factor(self, zam, e, S, Vy_extra, threshold=None):
```
Comment (luca-fiorito-11) [inline sandy/fy.py:891]: These will come in a different PR

File: sandy/fy.py [hunk 7]
```diff
@@ -593,84 +685,92 @@ def gls_cov_update(self, zam, e, Vy_extra=None,
...
-                   threshold=None):
+            Vy_extra = pd.DataFrame(Vy_extra)
+            S = S.reindex(index=Vy_extra.index).fillna(0)
+        Vx_post = Vx_prior.gls_update(S, Vy_extra=Vy_extra, rows=rows,
+                                          threshold=threshold).data.reindex(
+                                              index=index, columns=index)\
+                                              .fillna(0)
+        return sandy.CategoryCov(Vx_post)
+
+    def gls_update(self, zam, e, S, y_extra=None, Vy_extra=None,
```
Comment (luca-fiorito-11) [inline sandy/fy.py:767]: One or two outputs?

**Review by luca-fiorito-11 — CHANGES_REQUESTED** [review]
Can

**Comment by luca-fiorito-11** [comment]
@EnricaBelfiore , is this PR still relevant?

**Pull request closed without merging**